pants-plugins/examples/bash/shunit2_test_runner.py

```diff
@@ -0,0 +1,181 @@
+        ),
+    )
+    test_source_files_request = Get(
+        SourceFiles, SourceFilesRequest([request.field_set.sources])
+    )
+    dependencies_source_files, test_source_files = await MultiGet(
+        dependencies_source_files_request, test_source_files_request
+    )
+
+    # To check if the test files already have `source ./shunit2` in them, we need to look at the
+    # actual file content. We use `DigestContents` for this, and then use `CreateDigest` to create
+    # a digest of the (possibly) updated test files. See
+    # https://www.pantsbuild.org/v2.0/docs/rules-api-file-system.
+    #
+    # Most test runners don't modify their test files like we do here, so most test runners can
+    # skip this step.
+    test_files_content = await Get(
+        DigestContents, Digest, test_source_files.snapshot.digest
+    )
+    updated_test_files_content = []
+    for file_content in test_files_content:
+        if (
+            b"source ./shunit2" in file_content.content
+            or b". ./shunit2" in file_content.content
+        ):
+            updated_test_files_content.append(file_content)
+        else:
+            updated_file_content = FileContent(
+                path=file_content.path,
+                content=file_content.content + b"\nsource ./shunit2\n",
+            )
+            updated_test_files_content.append(updated_file_content)
+    updated_test_source_files = await Get(
+        Digest, CreateDigest(updated_test_files_content)
+    )
+
+    # The Process needs one single `Digest`, so we merge everything together. See
+    # https://www.pantsbuild.org/v2.0/docs/rules-api-file-system.
+    input_digest = await Get(
+        Digest,
+        MergeDigests(
+            [
+                shunit2_script,
+                updated_test_source_files,
+                dependencies_source_files.snapshot.digest,
+            ]
+        ),
+    )
+
+    process = Process(
+        argv=[bash_program.exe, *test_source_files.snapshot.files],
+        input_digest=input_digest,
+        description=f"Run shunit2 on {request.field_set.address}.",
+        level=LogLevel.DEBUG,
+        env=bash_setup.env_dict,
+    )
+    return TestSetup(process)
+
+
+@rule(desc="Run tests with Shunit2", level=LogLevel.DEBUG)
+async def run_tests_with_shunit2(field_set: Shunit2FieldSet) -> TestResult:
+    setup = await Get(TestSetup, TestSetupRequest(field_set))
+    # We use `FallibleProcessResult`, rather than `ProcessResult`, because we're okay with the
+    # Process failing.
+    result = await Get(FallibleProcessResult, Process, setup.process)
+    return TestResult.from_fallible_process_result(result)
+
+
+@rule(desc="Setup Shunit2 to run interactively", level=LogLevel.DEBUG)
+async def setup_shunit2_debug_test(field_set: Shunit2FieldSet) -> TestDebugRequest:
+    setup = await Get(TestSetup, TestSetupRequest(field_set))
+    # We set up an InteractiveProcess, which will get run in the `@goal_rule` in `test.py`.
```

Comment:
```suggestion
    # We set up an InteractiveProcess, which will be executed in the `@goal_rule` in `test.py`.
```
*extremely* subjective change, feel free to ignore

Reply:
I like it. Thanks.